scipy/io/_fast_matrix_market/meson.build

```diff
@@ -0,0 +1,33 @@
+subdir('fast_matrix_market/dependencies/fast_float')
+subdir('fast_matrix_market/dependencies/ryu')
+
+if meson.get_compiler('cpp').get_id() == 'msvc'
+  fmm_extra_args = ['/bigobj']
+else
+  fmm_extra_args = []
+endif
+
+py3.extension_module('_core',
+  [
+    'src/_core.hpp',
+    'src/_core.cpp',
+    'src/_core_read_array.cpp',
+    'src/_core_read_coo.cpp',
+    'src/_core_write_array.cpp',
+    'src/_core_write_coo.cpp',
+    'src/_core_write_csc.cpp',
+  ],
+  override_options: ['cpp_std=c++17'],
```

Comment:
I'm getting consistent build errors locally on a macOS arm64 machine in a conda-forge environment with Clang 11.1.0:
```
../scipy/io/_fast_matrix_market/fast_matrix_market/include/fast_matrix_market/fast_matrix_market.hpp:29:32: error: constexpr variable 'kSpace' must be initialized by a constant expression
    constexpr std::string_view kSpace = " ";
```
My first guess is that it has something to do with this override. It may be problematic to force C++17 like this, see gh-18408 for remaining issues with defaulting to it.

Reply:
Hmm. I just installed clang-11 from homebrew on my macOS arm64 machine and it compiles ok. I'll have to dig some more.

The problem in that issue is specific to `std::aligned_alloc` and I don't use that method. I'm also only enabling C++17 for this module and not globally, so everything else still builds as before. So far all the 64-bit tests pass on both macOS and Linux so the switch must be working. Hopefully no serious issues on Windows.

Reply:
@rgommers The cause is that when building with conda's clang-11 there is an extra `-std=c++14` argument added *after* the `-std=c++17` that my meson line adds. Of course the last `-std` argument is used. This extra argument does not appear for the Homebrew clang11, or the default clang.

Maybe someone familiar with the build process can shed some light on why, and also if all conda clang versions are similarly affected.

Here are the build commands in `build/compile_commands.json` for one of the FMM files for the two Clang 11 versions:

Conda clang-11 build command

```
clang++ -Iscipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p -Iscipy/io/_fast_matrix_market -I../scipy/io/_fast_matrix_market -I../scipy/io/_fast_matrix_market/fast_matrix_market/include -I../../../../../../opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev-clang11/lib/python3.10/site-packages/numpy/core/include -I../scipy/io/_fast_matrix_market/fast_matrix_market/dependencies/fast_float/include -Iscipy/io/_fast_matrix_market/fast_matrix_market/dependencies/ryu -I../scipy/io/_fast_matrix_market/fast_matrix_market/dependencies/ryu -I/opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev-clang11/include -I/opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev-clang11/include/python3.10 -fvisibility=hidden -fvisibility-inlines-hidden -fcolor-diagnostics -Wall -Winvalid-pch -std=c++17 -O2 -g -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -std=c++14 -isystem /opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev-clang11/include -D_FORTIFY_SOURCE=2 -isystem /opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev-clang11/include -DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION -DFMM_FROM_CHARS_INT_SUPPORTED -DFMM_TO_CHARS_INT_SUPPORTED -DFMM_USE_FAST_FLOAT -DFMM_USE_RYU -MD -MQ scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o -MF scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o.d -o scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o -c ../scipy/io/_fast_matrix_market/src/_core_write_csc.cpp
```


Homebrew clang-11 build command:

```
clang++ -Iscipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p -Iscipy/io/_fast_matrix_market -I../scipy/io/_fast_matrix_market -I../scipy/io/_fast_matrix_market/fast_matrix_market/include -I../../../../../../opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev/lib/python3.10/site-packages/numpy/core/include -I../scipy/io/_fast_matrix_market/fast_matrix_market/dependencies/fast_float/include -Iscipy/io/_fast_matrix_market/fast_matrix_market/dependencies/ryu -I../scipy/io/_fast_matrix_market/fast_matrix_market/dependencies/ryu -I/opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev/include -I/opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev/include/python3.10 -I/opt/homebrew/opt/llvm@11/include -fvisibility=hidden -fvisibility-inlines-hidden -fcolor-diagnostics -Wall -Winvalid-pch -std=c++17 -O2 -g -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /opt/homebrew/Caskroom/miniconda/base/envs/scipy-dev/include -DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION -DFMM_FROM_CHARS_INT_SUPPORTED -DFMM_TO_CHARS_INT_SUPPORTED -DFMM_USE_FAST_FLOAT -DFMM_USE_RYU -MD -MQ scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o -MF scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o.d -o scipy/io/_fast_matrix_market/_core.cpython-310-darwin.so.p/src__core_write_csc.cpp.o -c ../scipy/io/_fast_matrix_market/src/_core_write_csc.cpp
```

Reply:
This seems like it'll be a problem more broadly, so I opened a separate issue to track it: https://github.com/scipy/scipy/issues/18645